presto-main/src/main/java/com/facebook/presto/cost/CachingCostProvider.java

```diff
@@ -65,48 +78,221 @@ public CachingCostProvider(CostCalculator costCalculator, StatsProvider statsPro
         verify(!memo.getCumulativeCost(group).isPresent(), "Group cost already set");
         memo.storeCumulativeCost(group, cumulativeCost);
         return cumulativeCost;
     }
 
-    private PlanNodeCostEstimate calculateCumulativeCost(PlanNode node)
+    private PlanCostEstimate calculateCumulativeCost(PlanNode node)
     {
-        PlanNodeCostEstimate localCosts = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
+        PlanNodeCostEstimate localCost = costCalculator.calculateCost(node, statsProvider, lookup, session, types);
 
-        PlanNodeCostEstimate sourcesCost = sourceProvider.getSources(node).stream()
-                .map(this::getCumulativeCost)
-                .reduce(ZERO_COST, PlanNodeCostEstimate::add);
+        List<PlanCostEstimate> sourcesCosts = sourceProvider.getSources(node).stream()
+                .map(this::getPlanCost)
+                .collect(toImmutableList());
 
-        PlanNodeCostEstimate cumulativeCost = localCosts.add(sourcesCost);
-        return cumulativeCost;
+        MemoryEstimate memoryEstimate = node.accept(new PeakMemoryEstimator(), new EstimationContext(localCost.getMemoryCost(), sourcesCosts));
+        verify(memoryEstimate != null);
+
+        PlanCostEstimate sourcesCost = sourcesCosts.stream()
+                .reduce(PlanCostEstimate.zero(), CachingCostProvider::addSiblingsCost);
+
+        return new PlanCostEstimate(
+                localCost.getCpuCost() + sourcesCost.getCpuCost(),
+                Math.max(sourcesCost.getMaxMemory(), Math.max(memoryEstimate.maxMemoryBeforeOutputting, memoryEstimate.maxMemoryWhenOutputting)),
+                memoryEstimate.maxMemoryWhenOutputting,
+                localCost.getNetworkCost() + sourcesCost.getNetworkCost());
+    }
+
+    private static PlanCostEstimate addSiblingsCost(PlanCostEstimate a, PlanCostEstimate b)
+    {
+        return new PlanCostEstimate(
+                a.getCpuCost() + b.getCpuCost(),
+                a.getMaxMemory() + b.getMaxMemory(),
+                a.getMaxMemoryWhenOutputting() + b.getMaxMemoryWhenOutputting(),
+                a.getNetworkCost() + b.getNetworkCost());
+    }
+
+    private static class EstimationContext
+    {
+        private final double localMaxMemory;
+        private final List<PlanCostEstimate> sourcesCosts;
+
+        public EstimationContext(double localMaxMemory, List<PlanCostEstimate> sourcesCosts)
+        {
+            this.localMaxMemory = localMaxMemory;
+            this.sourcesCosts = requireNonNull(sourcesCosts, "sourcesCosts is null");
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .add("localMaxMemory", localMaxMemory)
+                    .add("sourcesCosts", sourcesCosts)
+                    .toString();
+        }
+    }
+
+    private static class MemoryEstimate
+    {
+        private final double maxMemoryBeforeOutputting;
```

Comment:
Add comments/javadoc about what all these things are.

Reply:
I would say something like "max memory allocation before node starts outputting". Would that be more helpful?